File: src/main/java/work/lclpnet/ap2/game/apocalypse_survival/util/AsSetup.java
package work.lclpnet.ap2.game.apocalypse_survival.util;

import net.minecraft.server.world.ServerWorld;
import org.json.JSONArray;
import org.json.JSONObject;
import work.lclpnet.ap2.impl.util.world.stage.Stage;
import work.lclpnet.ap2.impl.util.world.stage.StageReader;
import work.lclpnet.lobby.game.map.GameMap;

import java.util.LinkedList;
import java.util.List;
import java.util.Random;

public class AsSetup {

    private final GameMap map;
    private final ServerWorld world;
    private final Random random;

    public AsSetup(GameMap map, ServerWorld world, Random random) {
        this.map = map;
        this.world = world;
        this.random = random;
    }

    public List<MonsterSpawner> readSpawners() {
        List<MonsterSpawner> spawners = new LinkedList<>();

        JSONArray array = map.requireProperty("spawners");

        for (Object o : array) {
            if (!(o instanceof JSONObject json)) continue;

            spawners.add(createSpawner(json));
        }

        return spawners;
    }

    private MonsterSpawner createSpawner(JSONObject json) {
        JSONObject stageJson = json.getJSONObject("stage");

        Stage stage = StageReader.readStage(stageJson);

        return new MonsterSpawner(world, stage, random);
    }
}

File: src/minigames/apocalypse_survival/src/main/kotlin/work/lclpnet/ap2/game/apocalypse_survival/util/AsSetup.kt
package work.lclpnet.ap2.game.apocalypse_survival.util

import net.minecraft.server.level.ServerLevel
import org.json.JSONArray
import org.json.JSONObject
import work.lclpnet.ap2.impl.map.MapUtil
import work.lclpnet.game.map.GameMap
import java.util.*

class AsSetup(
    private val map: GameMap,
    private val world: ServerLevel,
    private val random: Random,
    private val targetManager: TargetManager
) {

    fun readSpawners(): List<MonsterSpawner> {
        val array = map.requireProperty<JSONArray>("spawners")

        return array.filterIsInstance<JSONObject>().map { createSpawner(it) }
    }

    private fun createSpawner(json: JSONObject): MonsterSpawner {
        val stageJson = json.getJSONObject("stage")
        val blockShape = MapUtil.readShape(stageJson)

        val minSpawnTicks = json.getNumber("min-spawn-ticks").toInt()
        val maxSpawnTicks = json.getNumber("max-spawn-ticks").toInt()

        return MonsterSpawner(world, blockShape, random, targetManager, minSpawnTicks, maxSpawnTicks)
    }
}
